src/allocation_engine/README.md

```diff
@@ -0,0 +1,135 @@
+pub(crate) struct InnerNode {
+    /// Interval handled by this node.
+    pub(crate) key: Range,
+    /// Optional contained data, None if the node is free.
+    pub(crate) node_state: NodeState,
+    /// Optional left child of current node.
+    pub(crate) left: Option<Box<InnerNode>>,
+    /// Optional right child of current node.
+    pub(crate) right: Option<Box<InnerNode>>,
+    /// Cached height of the node.
+    pub(crate) height: u64,
+    /// Cached maximum valued covered by this node.
+    pub(crate) max_key: u64,
+}
+
+pub enum NodeState {
+    /// Node is free
+    Free,
+    /// Node is allocated but without associated data
+    Allocated,
+}
+
+/// Insert the ket into the interval tree, returns Error if intersects with existing nodes.
+pub fn insert(&mut self, key: Range, node_state: NodeState) -> Result<()> { }
+
+/// Update an existing entry and return the old value.
+pub(crate) fn update(&mut self, key: &Range, node_state: NodeState) -> Result<()> { }
+
+/// Remove the `key` from the tree and return the associated data.
+pub fn delete(&mut self, key: &Range) -> Option<()> { }
+
+ /// Construct a new empty IntervalTree.
+pub fn new() -> Self { }
+
+/// Check whether the interval tree is empty.
+pub fn is_empty(&self) -> bool { }
+
+/// Get the data item associated with the key, or return None if no match found.
+pub fn get(&self, key: &Range) -> Option<NodeState> { }
+
+/// Get a shared reference to the node fully covering the entire key range.
+pub fn get_superset(&self, key: &Range) -> Option<(&Range, NodeState)> { }
+```
+
+### Address allocation policy
+
+We offer three options for placing and memory slot in the managed address space:
+
+1. `LastMatch` -> When using this allocation policy the allocator will try to
+insert the range described by the constraint at the first available position
+starting from the end of the managed address space.
+2. `FirstMatch` -> When using this allocation policy the allocator will try to
+insert the range described by the constraint at the first available position
+starting from the begining of the managed address space.
+3. `ExactMatch` -> When using this allocation policy the allocator will try to
+insert the range at the exact position described by the constraint, otherwise
+it will return an error.
+
+## Usage
+
+The concept of Interval Tree may seem complicated, but using vm-allocator to do
+resource allocation and release is simple and straightforward.
+
+```rust
+// 1. To start with, we should create an interval tree for some specific resouces 
+// and give an address space as a root node
+
+#[derive(Debug, PartialEq)]
+pub struct AddressAllocator {
+    base: u64,
+    end: u64,
+    interval_tree: IntervalTree,
+}
+
+impl AddressAllocator {
+    pub fn new(base: u64, size: u64) -> std::result::Result<Self, Error> {
+        let interval_tree = IntervalTree::new();
+        let mut address_allocator = AddressAllocator {
+            base: base,
+            end: end.unwrap(),
```

Comment:
Where does `end` come from and why `.unwrap()`? Same question for the `allocate()` below.